autopush/senderids.py

```diff
@@ -58,42 +58,44 @@ def __init__(self, args):
                 log.err("senderid_list is not a dict. Ignoring")
             else:
                 self.update(senderIDs)
-        self._refresh()
+        # set up auto-refresh for senderIDs
+        if self._use_s3:
+            self.service = TimerService(self._expry, self._refresh)
+            self.service.startService()
 
-    def _write(self, bucket, senderIDs):
+    def _write(self, senderIDs, *args):
         """Write a list of SenderIDs to S3"""
+        bucket = self.conn.get_bucket(self.ID)
         key = Key(bucket)
         key.key = self.KEYNAME
         key.set_contents_from_string(json.dumps(senderIDs))
         self._senderIDs = senderIDs
-        self._updated = time.time()
 
-    def _create(self, senderIDs):
+    def _create(self, *args):
         """Create a new bucket containing the senderIDs"""
-        bucket = self.conn.create_bucket(self.ID)
-        self._write(bucket, senderIDs)
+        self.conn.create_bucket(self.ID)
+        self._write(self._senderIDs)
+
+    def _update_senderIDs(self, *args):
+        bucket = self.conn.get_bucket(self.ID)
+        key = Key(bucket)
+        key.key = self.KEYNAME
+        candidates = json.loads(key.get_contents_as_string())
+        if candidates:
+            if type(candidates) is not dict:
+                log.err("Wrong data type stored for senderIDs. "
+                        "Should be dict. Ignoring.")
+                return
+            self._senderIDs = candidates
 
     def _refresh(self):
         """Refresh the senderIDs from the S3 bucket"""
         if not self._use_s3:
             return
-        # Only refresh if needed.
-        if time.time() < self._updated + self._expry:
-            return
-        try:
-            bucket = self.conn.get_bucket(self.ID)
-            key = Key(bucket)
-            key.key = self.KEYNAME
-            candidates = json.loads(key.get_contents_as_string())
-            if candidates:
-                if type(candidates) is not dict:
-                    log.err("Wrong data type stored for senderIDs. "
-                            "Should be dict. Ignoring.")
-                    return
-                self._senderIDs = candidates
-                self._updated = time.time()
-        except S3ResponseError:
-            self._create(self._senderIDs)
+        d = Deferred()
+        d.addCallback(self._update_senderIDs)
```

Comment:
The function still blocks in the loop, you want to use deferToThread which returns a deferred, but kicks the function to run in a thread. You also want _create to run in the thread with it.